linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/SJSGen.scala

```diff
@@ -145,6 +145,38 @@ private[emitter] final class SJSGen(
     val newArrayOfThisClass = "newArrayOfThisClass"
   }
 
+  /* This is a `val` because it is used at the top of every file, outside of
+   * any cache. Fortunately it does not depend on any dynamic content.
+   */
+  val declarePrototypeVar: List[Tree] = {
+    implicit val pos = Position.NoPosition
+    if (minify) VarDef(fileLevelVarIdent(VarField.p), None) :: Nil
+    else Nil
+  }
+
+  def prototypeFor(classRef: Tree)(implicit pos: Position): Tree = {
+    import TreeDSL._
+    if (minify) fileLevelVar(VarField.p)
+    else classRef.prototype
+  }
+
+  def genAssignPrototype(classRef: Tree, value: Tree, localDecl: Boolean = false)(implicit pos: Position): Tree = {
+    import TreeDSL._
+    val assign = classRef.prototype := value
+    if (!minify)
+      assign
+    else if (localDecl)
+      VarDef(fileLevelVarIdent(VarField.p), Some(assign))
+    else
+      fileLevelVar(VarField.p) := assign
+  }
+
+  def setPrototypeVar(rhs: Tree)(implicit pos: Position): List[Tree] = {
```

Comment:
> I have just noticed this: This is always called in the following form `setPrototypeVar(<x>.prototype)`.
> 
> Obviously, that's a good thing. But maybe we should move the `.prototype` into this method?

:+1:

> Reading through this, I'm actually wondering whether we should not make `genAssignPrototype` handle a missing value: it always corresponds to the super class. IIUC this would also simplify the codepath of `chainProtoWithGlobals` in `ClassEmitter`.

I don't see what would improve with that change. `chainProtoWithGlobals` would avoid one branch, but at the cost of having one more branch in `genAssignPrototype`. And it's the only call site that that calls both `genAssignPrototype` and `setPrototypeVar`. All the other ones use or the other statically.